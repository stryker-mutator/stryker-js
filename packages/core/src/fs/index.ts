export * from './file-system.js';
export * from './project.js';
export * from './project-file.js';
export * from './project-reader.js';
