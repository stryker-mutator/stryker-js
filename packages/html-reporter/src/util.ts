import * as path from 'path';
import { promisify } from 'util';

import { fsAsPromised } from '@stryker-mutator/util';
import * as mkdirp from 'mkdirp';
import * as rimraf from 'rimraf';

export function copyFile(fromFilename: string, toFilename: string): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const readStream = fsAsPromised.createReadStream(fromFilename);
    const writeStream = fsAsPromised.createWriteStream(toFilename);
    readStream.on('error', reject);
    writeStream.on('error', reject);
    readStream.pipe(writeStream);
    readStream.on('end', resolve);
  });
}

export const deleteDir = promisify(rimraf);

export const mkdir = promisify(mkdirp);

export async function writeFile(fileName: string, content: string) {
  await mkdir(path.dirname(fileName));
  await fsAsPromised.writeFile(fileName, content, 'utf8');
}
