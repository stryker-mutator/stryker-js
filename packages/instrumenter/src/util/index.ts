export * from './syntax-helpers.js';
export * from './binary-operator.js';
