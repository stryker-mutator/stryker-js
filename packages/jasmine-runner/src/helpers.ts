import { TestResult, TestStatus } from '@stryker-mutator/api/test-runner';
import jasmine from 'jasmine';

export const helpers = {
  toStrykerTestResult(specResult: jasmine.SpecResult, timeSpentMs: number): TestResult {
    const baseResult = {
      id: specResult.id.toString(),
      name: specResult.fullName,
      timeSpentMs,
    };
    if (specResult.status === 'disabled' || specResult.status === 'pending' || specResult.status === 'excluded') {
      return {
        ...baseResult,
        status: TestStatus.Skipped,
      };
    } else if (!specResult.failedExpectations || specResult.failedExpectations.length === 0) {
      return {
        ...baseResult,
        status: TestStatus.Success,
      };
    } else {
      return {
        ...baseResult,
        status: TestStatus.Failed,
        failureMessage: specResult.failedExpectations.map((failedExpectation) => failedExpectation.message).join(','),
      };
    }
  },
  createJasmine(options: jasmine.JasmineOptions): jasmine {
    return new jasmine(options);
  },
};
