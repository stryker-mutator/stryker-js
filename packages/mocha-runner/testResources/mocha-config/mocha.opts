--async-only
--blaat true
--spec /foo/*.js
/tests/**/*.js