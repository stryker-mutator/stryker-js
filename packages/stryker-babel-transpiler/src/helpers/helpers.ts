import * as path from 'path';

export function toJSFileName(fileName: string) {
  const fileNameInfo = path.parse(fileName);
  const newFileName = path.join(fileNameInfo.dir, fileNameInfo.name + '.js');
  return newFileName;
}
