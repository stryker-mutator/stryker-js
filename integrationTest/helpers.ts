import * as path from 'path';
import * as fs from 'mz/fs';
import { expect } from 'chai';
import { ScoreResult } from 'stryker-api/report';

export async function readScoreResult(eventResultDirectory = path.resolve('reports', 'mutation', 'events')) {
    const allReportFiles = await fs.readdir(eventResultDirectory);
    const scoreResultReportFile = allReportFiles.find(file => !!file.match(/.*onScoreCalculated.*/));
    expect(scoreResultReportFile).ok;
    const scoreResultContent = await fs.readFile(path.resolve(eventResultDirectory, scoreResultReportFile || ''), 'utf8');
    return JSON.parse(scoreResultContent) as ScoreResult;
}