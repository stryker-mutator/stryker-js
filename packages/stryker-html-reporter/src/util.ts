import { fsAsPromised } from '@stryker-mutator/util';
import * as path from 'path';
import * as mkdirp from 'mkdirp';
import * as rimraf from 'rimraf';

async function copyFolderOrFile(fromPath: string, toPath: string): Promise<void> {
  const stats = await fsAsPromised.stat(fromPath);
  if (stats.isDirectory()) {
    return copyFolder(fromPath, toPath);
  }
  else {
    return copyFile(fromPath, toPath);
  }
}

export function copyFolder(fromPath: string, to: string): Promise<void> {
  return mkdir(to)
    .then(() => fsAsPromised.readdir(fromPath))
    .then(files => Promise.all(files.map(file => copyFolderOrFile(path.join(fromPath, file), path.join(to, file)))))
    .then(_ => void 0);
}

function copyFile(fromFilename: string, toFilename: string): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    const readStream = fsAsPromised.createReadStream(fromFilename);
    const writeStream = fsAsPromised.createWriteStream(toFilename);
    readStream.on('error', reject);
    writeStream.on('error', reject);
    readStream.pipe(writeStream);
    readStream.on('end', () => resolve());
  });
}

export function deleteDir(dirToDelete: string): Promise<void> {
  return new Promise((res, rej) => {
    rimraf(dirToDelete, err => {
      if (err) {
        rej(err);
      } else {
        res();
      }
    });
  });
}

export function mkdir(folderName: string): Promise<void> {
  return new Promise<void>((res, rej) => {
    mkdirp(folderName, err => {
      if (err) {
        rej(err);
      } else {
        res();
      }
    });
  });
}

export function writeFile(fileName: string, content: string) {
  return mkdir(path.dirname(fileName))
    .then(_ => fsAsPromised.writeFile(fileName, content, 'utf8'));
}

export function countPathSep(fileName: string) {
  let count = 0;
  for (const ch of fileName) {
    if (ch === path.sep) {
      count++;
    }
  }
  return count;
}
