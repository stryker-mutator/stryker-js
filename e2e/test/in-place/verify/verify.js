import { promises as fsPromises } from 'fs';
import path from 'path';
import { promisify } from 'util';
import { fileURLToPath, URL } from 'url';

import chai from 'chai';
import { execa, execaSync } from 'execa';
import rimraf from 'rimraf';
import { it } from 'mocha';
import chaiAsPromised from 'chai-as-promised';

import { expectMetricsJsonToMatchSnapshot } from '../../../helpers.js';

chai.use(chaiAsPromised);
const expect = chai.expect;
const rm = promisify(rimraf);

const rootResolve = path.resolve.bind(path, fileURLToPath(new URL('..', import.meta.url)));

describe('in place', () => {
  /**
   * @type {string}
   */
  let originalAddJSContent;
  function readAddJS() {
    return fsPromises.readFile(rootResolve('src', 'Add.js'), 'utf-8');
  }
  before(async () => {
    originalAddJSContent = await readAddJS();
  });
  afterEach(async () => {
    await rm(rootResolve('reports'));
    await rm(rootResolve('.lock'));
  });
  it('should reset files after a successful run', async () => {
    execaSync('stryker', ['run']);
    const addJSContent = await fsPromises.readFile(rootResolve('src', 'Add.js'), 'utf-8');
    expect(addJSContent).eq(originalAddJSContent);
  });
  it('should report correct score', async () => {
    execaSync('stryker', ['run']);
    await expectMetricsJsonToMatchSnapshot();
  });
  it('should also reset the files if Stryker exits unexpectedly', async () => {
    // Arrange
    let addJSMutatedContent;
    await fsPromises.writeFile(rootResolve('.lock'), ''); // this will lock the test run completion
    const onGoingStrykerRun = execa('node', [path.resolve('..', '..', 'node_modules', '.bin', 'stryker'), 'run']);
    onGoingStrykerRun.stdout.on('data', async (data) => {
      if (data.toString().includes('Starting initial test run')) {
        addJSMutatedContent = await readAddJS();
        // Now, mr bond, it is time to die!
        onGoingStrykerRun.kill();
      }
    });
    // Act
    await expect(onGoingStrykerRun).rejected;
    // Assert
    expect(await readAddJS()).eq(originalAddJSContent);
    expect(addJSMutatedContent).not.eq(originalAddJSContent);
  });
});
