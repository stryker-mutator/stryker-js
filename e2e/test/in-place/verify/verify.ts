import { expectMetrics } from '../../../helpers';
import { promises as fsPromises } from 'fs';
import chai from 'chai';
import execa from 'execa';
import rimraf from 'rimraf';
import path from 'path';
import { promisify } from 'util';
import { it } from 'mocha';
import chaiAsPromised from 'chai-as-promised';

chai.use(chaiAsPromised);
const expect = chai.expect;

const rm = promisify(rimraf);

const rootResolve: typeof path.resolve = path.resolve.bind(path, __dirname, '..');

describe('in place', () => {

  let originalAddJSContent: string;


  function readAddJS(): Promise<string> {
    return fsPromises.readFile(rootResolve('src', 'Add.js'), 'utf-8');
  }

  before(async () => {
    originalAddJSContent = await readAddJS();
  })


  afterEach(async () => {
    await rm(rootResolve('reports'));
    await rm(rootResolve('.lock'));
  })

  it('should reset files after a successful run', async () => {
    execa.sync('stryker', ['run']);
    const addJSContent = await fsPromises.readFile(rootResolve('src', 'Add.js'), 'utf-8');
    expect(addJSContent).eq(originalAddJSContent);
  });

  it('should report correct score', async () => {
    execa.sync('stryker', ['run']);
    await expectMetrics({ mutationScore: 73.68 });
  });

  it('should also reset the files if Stryker exits unexpectedly', async () => {
    // Arrange
    let addJSMutatedContent: string;
    await fsPromises.writeFile(rootResolve('.lock'), ''); // this will lock the test run completion
    const onGoingStrykerRun = execa('node', [path.resolve('..', '..', 'node_modules', '.bin', 'stryker'), 'run']);
    onGoingStrykerRun.stdout.on('data', async (data) => {
      if (data.toString().includes('Starting initial test run')) {
        addJSMutatedContent = await readAddJS();

        // Now, mr bond, it is time to die!
        onGoingStrykerRun.kill();
      }
    });

    // Act
    await expect(onGoingStrykerRun).rejected;

    // Assert
    expect(await readAddJS()).eq(originalAddJSContent);
    expect(addJSMutatedContent).not.eq(originalAddJSContent);
  });
});
