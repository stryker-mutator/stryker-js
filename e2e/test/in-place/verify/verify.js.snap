// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in place should report correct score 1`] = `
Object {
  "compileErrors": 0,
  "ignored": 0,
  "killed": 14,
  "mutationScore": 73.68421052631578,
  "mutationScoreBasedOnCoveredCode": 100,
  "noCoverage": 5,
  "runtimeErrors": 0,
  "survived": 0,
  "timeout": 0,
  "totalCovered": 14,
  "totalDetected": 14,
  "totalInvalid": 0,
  "totalMutants": 19,
  "totalUndetected": 5,
  "totalValid": 19,
}
`;
